Sources/Core/Microsoft.StreamProcessing/Aggregates/SlidingMaxAggregate.cs

```diff
@@ -44,51 +52,90 @@ private MinMaxState<T> Accumulate(MinMaxState<T> state, long timestamp, T input)
             }
             else
             {
-                if (state.currentTimestamp != InvalidSyncTime)
-                    state.savedValues.Add(state.currentValue);
+                // save only if new input is smaller, otherwise overwrite current with new
+                if (state.currentTimestamp != InvalidSyncTime &&
+                    this.comparer(input, state.currentValue) < 0)
+                {
+                    PushToCollection(state);
+                }
                 state.currentTimestamp = timestamp;
                 state.currentValue = input;
             }
             return state;
         }
 
+        private void PushToCollection(MinMaxState<T> state)
+        {
+            Contract.Assert(state.currentTimestamp != InvalidSyncTime);
+            Contract.Assert(state.values.Count == 0 || state.values.Last().timestamp <= state.currentTimestamp); // Ensure new timestamp is higher
+
+            int index = state.values.FindLastIndex(e => this.comparer(e.value, state.currentValue) > 0);
+            if (index == -1)
+                state.values.Clear();
+            else
+                state.values.RemoveRange(index + 1, state.values.Count - index - 1);
+
+            state.values.Add(new ValueAndTimestamp<T> { timestamp = state.currentTimestamp, value = state.currentValue });
+            state.currentTimestamp = InvalidSyncTime;
```

Comment:
>state.currentTimestamp = InvalidSyncTime; [](start = 12, length = 41)

This seems out of place in this function. Is this necessary?

Reply:
it is not obvious that state would be modified by this function, but taking a step back, it doesn't need to be, since the only caller of this function immediately overwrites state.currentTimestamp anyway, right?

---
In reply to: [349384811](https://github.com/microsoft/Trill/pull/119#discussion_r349384811) [](ancestors = 349384811)